### ISSUE TITLE
PHP Notice: Undefined property: stdClass::$error in wp-rocket\inc\Addon\Cloudflare\Cloudflare.php on line 131

### ISSUE THREAD

**Issue by arunbasillal** [wp-media/wp-rocket#2369]
**Describe the bug**
Saving Cloudflare credentials on Local Lightning 5.2.5 generates the following error. 

```
PHP Notice:  Undefined property: stdClass::$errors in C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-content\plugins\wp-rocket\inc\Addon\Cloudflare\Cloudflare.php on line 131
PHP Stack trace:
PHP   1. {main}() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-admin\options.php:0
PHP   2. update_option() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-admin\options.php:285
PHP   3. do_action() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\option.php:424
PHP   4. WP_Hook->do_action() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\plugin.php:478
PHP   5. WP_Hook->apply_filters() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\class-wp-hook.php:312
PHP   6. WPMedia\Cloudflare\Subscriber->save_cloudflare_options() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\class-wp-hook.php:290
PHP   7. WPMedia\Cloudflare\Cloudflare->is_api_keys_valid() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-content\plugins\wp-rocket\inc\Addon\Cloudflare\Subscriber.php:371

PHP Warning:  Invalid argument supplied for foreach() in C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-content\plugins\wp-rocket\inc\Addon\Cloudflare\Cloudflare.php on line 131
PHP Stack trace:
PHP   1. {main}() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-admin\options.php:0
PHP   2. update_option() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-admin\options.php:285
PHP   3. do_action() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\option.php:424
PHP   4. WP_Hook->do_action() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\plugin.php:478
PHP   5. WP_Hook->apply_filters() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\class-wp-hook.php:312
PHP   6. WPMedia\Cloudflare\Subscriber->save_cloudflare_options() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\class-wp-hook.php:290
PHP   7. WPMedia\Cloudflare\Cloudflare->is_api_keys_valid() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-content\plugins\wp-rocket\inc\Addon\Cloudflare\Subscriber.php:371

PHP Warning:  Cannot modify header information - headers already sent by (output started at C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-content\plugins\wp-rocket\inc\Addon\Cloudflare\Cloudflare.php:131) in C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\pluggable.php on line 1265
PHP Stack trace:
PHP   1. {main}() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-admin\options.php:0
PHP   2. wp_redirect() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-admin\options.php:313
PHP   3. header() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\pluggable.php:1265

PHP Warning:  Cannot modify header information - headers already sent by (output started at C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-content\plugins\wp-rocket\inc\Addon\Cloudflare\Cloudflare.php:131) in C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\pluggable.php on line 1268
PHP Stack trace:
PHP   1. {main}() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-admin\options.php:0
PHP   2. wp_redirect() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-admin\options.php:313
PHP   3. header() C:\Users\Arun Basil Lal\Local Sites\a-test-site\app\public\wp-includes\pluggable.php:1268
```

Also here: http://snippi.com/raw/q5c4b7p

**To Reproduce**
1. Enable Cloudflare add-on on Local Lightning (the new one). 
2. Enter credentials and save it. 
3. You should see the error. 
4. If not, clear ZONE ID. Save settings.
5. Enter ZONE ID again. Save settings. 
6. See error. 

**Screenshots**
Please refer this for an example of how to replicate this - https://wp-media.slack.com/archives/CBJ2KQ0MU/p1583264526169400?thread_ts=1583259272.164000&cid=CBJ2KQ0MU
(Contains sensitive data)

**Additional context**
- Related conversation - https://wp-media.slack.com/archives/CBJ2KQ0MU/p1583259272164000

**Comment by hellofromtonya** [comment]
More details for this particular error:

- It's not isolated to Cloudflare. Rather, it's a cURL + SSL verify problem in the localhost.
- This problem occurs within Local Lightning and Local by Flywheel. 
- Can't reproduce with my other localhost tools. Only with Local.

**How to reproduce the problem?**

1. Make sure you are using Local Lightning or Local by Flywheel.
2. [Add this file](https://gist.github.com/hellofromtonya/4d7c2eb85378231110e045f409be9d3e) into your `public/wp-content/mu-plugins/`.
3. Open the local website in your browser.

Here is what I get in the browser from the `var_dump`s.

```
/app/public/wp-content/mu-plugins/dev-overrides.php:18:
object(WP_Error)[331]
  public 'errors' => 
    array (size=1)
      'http_request_failed' => 
        array (size=1)
          0 => string 'cURL error 60: SSL certificate problem: self signed certificate' (length=63)
  public 'error_data' => 
    array (size=0)
      empty
/app/public/wp-content/mu-plugins/dev-overrides.php:19:string '' (length=0)
/app/public/wp-content/mu-plugins/dev-overrides.php:20:string '' (length=0)
```

**Exploring it deeper**

Notice the `cURL error 60` from above. If you change `sslverify` to `false`, it works. No errors.

**Why does Arun get PHP Notice for Cloudflare**

The Cloudflare Addon expects to get back an error jSON response from Cloudflare. Instead, when running `get_zones()` on Local localhost, it gets this curl error 60 back as a string. The format is not what is expected as it never communicated outside of the localhost, i.e. never made the remote call to Cloudflare.

**Comment by hellofromtonya** [comment]
If we seek to fix this far edgy case problem, then I suggest that we audit all of our remote requests (i.e. whether using cURL directly or through the WP HTTP API). We want to validate that we are handling non-HTTP response errors too, such as the `cURL error 60` error.

**Comment by crystinutzaa** [comment]
For the specific Cloudflare, there is a fix which is really easy to apply. In the file:
`inc/Addon/Cloudflare/APIClient.php`, line 317 need to be modified like this (to keep Cloudflare error reply format):
```
$errors                    = new stdClass();
$errors->code       = $http_code;
$errors->message  = $error;
$response->errors  = [ $errors ];

**Comment by hellofromtonya** [comment]
**Reproduce the issue** ✅

Yes, I can reproduce this issue by following the steps I outlines above.

**Identify the root cause** ✅

The actual root cause is we are only handling Cloudflare errors in the zone check and not cURL errors.

What do I mean?

- Cloudflare stores its errors in the `errors` property as an array.
- A cURL error is stored in the `error` property. 

2 different properties.

Notice how we are handling it below:

```php
public function is_api_keys_valid( $cf_email, $cf_api_key, $cf_zone_id ) {
	// removed the code here as it's not needed to show the problem.
	try {
		$this->api->set_api_credentials( $cf_email, $cf_api_key, $cf_zone_id );

		$cf_zone = $this->api->get_zones();

		if ( empty( $cf_zone->success ) ) {
			foreach ( $cf_zone->errors as $error ) {
```

**Scope a solution** ✅

For this specific problem, in the `Cloudflare\Cloudflare::is_api_keys_valid()` method (shown above), we need to handle the `error` property for cURL errors. 

That means adding code in this method to throw the correct `WP_Error` with a message. 

Why not modify the `ClientAPI` to add the cURL error to the `errors` array? 

A couple of reasons:
1. It's not a Cloudflare error. It's a cURL error. It can help us if we ever need to know one from the other.
2. If a cURL error happens, it does not mean the Zone ID is incorrect per se. Rather, it can be something else. By keeping it separate, we can simplify our error handler code to throw the right `WP_Error` message.

Here's a prototype of it:

```php
public function is_api_keys_valid( $cf_email, $cf_api_key, $cf_zone_id ) {
	// removed the code here as it's not needed to show the problem.
	try {
		$this->api->set_api_credentials( $cf_email, $cf_api_key, $cf_zone_id );

		$cf_zone = $this->api->get_zones();

		if ( empty( $cf_zone->success ) ) {

			// cURL error exists.
			if ( $cf_zone->error ) {
				return new WP_Error( 'cloudflare_invalid_auth', $cf_zone->error );
			}

			// Cloudflare error(s) exists.
			foreach ( $cf_zone->errors as $error ) {
```

**Estimate the effort** ✅

`Effort: [S]`

**Comment by hellofromtonya** [comment]
@crystinutzaa What do you think?

**Comment by crystinutzaa** [comment]
@hellofromtonya which is the prefered solution? I would go with Solution 1 which is more performant.

**Comment by crystinutzaa** [comment]
The PR is available for review in here:
https://github.com/wp-media/cloudflare/pull/21

**Comment by hellofromtonya** [comment]
@crystinutzaa Approved and merged in the Cloudflare repo. The new version of Cloudflare needs to be loaded into Rocket. Please create a new PR. We'll do one more QA check on the completed version. Then get it merged as part of 3.5.2 release.
